### PR TITLE
feat: this.emitFile support prebuilt-chunk type

### LINKED ISSUE #4210: Provide an way to modify Emited chunk in generateBundle | writeBundle hooks
### Feature Use Case

Given that rollup's bundle speedy is not fast, I want to leave third party module bundle to esbuild(and rollup bundle user code), considering supports code splitting, I need to modify the emited chunk in generateBundle hooks, which seem not supported yet
```ts
    async generateBundle(options, chunkMap) {
      const entryPoints = {};
      for (const [file, chunk] of Object.entries(chunkMap)) {
        if (file.endsWith('.js')) {
          entryPoints[path.resolve(config.root, file)] = chunk;
        }
      }
      const context = this;
      const result = await build({
        entryPoints: Object.keys(entryPoints),
        write: false,
        absWorkingDir: __dirname,
        bundle: true,
        outdir: '/assets',
        format: 'esm',
        allowOverwrite:true,
        splitting:true,
        plugins: [
          {
            name: 'memfs',
            setup(build) {
              build.onResolve({ filter: /.*/ },async (args) => {
                const realPath = path.resolve(args.resolveDir, args.path);
                if (entryPoints[realPath]) {
                  return {
                    path: realPath
                  };
                }else {
                  const result = await context.resolve(args.path, args.importer, {
                    skipSelf:true
                  })
                  return {
                    path: result.id
                  }
                }
              });
              build.onLoad({ filter: /.*/ }, (args) => {
                const res = entryPoints[args.path];
                if (res) {
                  return {
                    contents: entryPoints[args.path].code,
                    resolveDir: path.dirname(args.path)
                  };
                }
              });
            }
          }
        ]
      });
      for(const output of result.outputFiles){
        const fileName= path.relative('/', output.path);
        console.log('fileName:', fileName);
        // rollup not support emitChunk in generate bundle,which is hard to deal with
        this.emitFile({
          type: 'chunk', // support modify 'asset' but not support modify 'chunk'
          fileName: fileName,
          source: output.text
        })
      }
    }
```

### Feature Proposal

provide an hook likes `updateFile` to modify emited chunk & asset

### PR DESCRIPTION
**Pull request by TrickyPi**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:
resolves #4210 

### Description
This PR implements a basic feature about emitting prebuilt chunks, outlined in https://github.com/rollup/rollup/issues/4210#issuecomment-1519387621

### CODE REVIEW

File: src/utils/FileEmitter.ts [hunk 1]
```diff
@@ -367,6 +404,49 @@ export class FileEmitter {
 		return this.assignReferenceId(consumedChunk, emittedChunk.id);
 	}
 
+	private emitPrebuiltChunk(
+		emitPrebuiltChunk: Omit<EmittedFile, 'fileName' | 'name'> &
+			Pick<EmittedPrebuiltChunk, 'exports' | 'map'>
+	): string {
+		if (typeof emitPrebuiltChunk.code !== 'string') {
```
Comment (lukastaegert) [inline src/utils/FileEmitter.ts:411]: Both validations should be covered in separate `function` tests, cf. e.g. `test/function/samples/emit-file/invalid-asset-name` for where to place them and what they could look like.

File: src/utils/FileEmitter.ts [hunk 2]
```diff
@@ -367,6 +404,49 @@ export class FileEmitter {
...
+			map: emitPrebuiltChunk.map,
+			referenceId: '',
+			type: 'prebuilt-chunk'
+		};
+		const referenceId = this.assignReferenceId(
+			consumedPrebuiltChunk,
+			consumedPrebuiltChunk.fileName
+		);
+		if (this.output) {
+			this.output.bundle[consumedPrebuiltChunk.fileName] =
```
Comment (lukastaegert) [inline src/utils/FileEmitter.ts:444]: This scenario is where a plugin emits a prebuilt-chunk during output generation. To test this, e.g. emit such a chunk in `generateBundle`.

**Review by lukastaegert — CHANGES_REQUESTED** [review]
Looks good! However, we need some more tests, and then we need to update the documentation.

File: src/utils/FileEmitter.ts [hunk 3]
```diff
@@ -367,6 +404,49 @@ export class FileEmitter {
...
+					`Emitted prebuilt chunks need to have a valid string code, received "${emitPrebuiltChunk.code}"`
+				)
+			);
+		}
+		if (
+			typeof emitPrebuiltChunk.fileName !== 'string' ||
+			!(
+				typeof emitPrebuiltChunk.fileName === 'string' &&
+				!isPathFragment(emitPrebuiltChunk.fileName)
+			)
```
[the remark this answers is not in the document]
Comment (TrickyPi) [reply src/utils/FileEmitter.ts]: awesome spot!

File: src/utils/FileEmitter.ts [hunk 4]
```diff
@@ -92,25 +98,26 @@ interface EmittedFile {
...
+function hasValidType(emittedFile: unknown): emittedFile is {
+	[key: string]: unknown;
+	type: 'asset' | 'chunk' | 'prebuilt-chunk';
+} {
 	return Boolean(
 		emittedFile &&
 			((emittedFile as { [key: string]: unknown }).type === 'asset' ||
-				(emittedFile as { [key: string]: unknown }).type === 'chunk')
+				(emittedFile as { [key: string]: unknown }).type === 'chunk' ||
+				(emittedFile as { [key: string]: unknown }).type === 'prebuilt-chunk')
```
[the remark this answers is not in the document]
Comment (TrickyPi) [reply src/utils/FileEmitter.ts]: done

File: src/utils/FileEmitter.ts [hunk 5]
```diff
@@ -216,6 +226,9 @@ export class FileEmitter {
 		if (emittedFile.type === 'chunk') {
 			return getChunkFileName(emittedFile, this.facadeChunkByModule);
 		}
+		if (emittedFile.type === 'prebuilt-chunk') {
```
[the remark this answers is not in the document]
Comment (TrickyPi) [reply src/utils/FileEmitter.ts:228]: done

File: src/utils/FileEmitter.ts (hunk 2 above)
Comment (TrickyPi) [reply src/utils/FileEmitter.ts:440]: done

File: src/utils/FileEmitter.ts [hunk 6]
```diff
@@ -188,6 +195,9 @@ export class FileEmitter {
 				)
```
[the remark this answers is not in the document]
Comment (TrickyPi) [reply src/utils/FileEmitter.ts:194]: done

**Comment by TrickyPi** [comment]
docs has been updated.

**Review by lukastaegert — APPROVED** [review]
Nice! I slightly changed the wording of the docs, but overall I will release this now.

**Pull request merged**